package/kedro_viz/integrations/kedro/hooks_utils.py

```diff
@@ -0,0 +1,132 @@
+"""Utility functions for Kedro hooks implementation."""
+
+import json
+import logging
+from datetime import datetime, timezone
+from pathlib import Path
+from typing import Any, Dict, List, Optional
+
+import fsspec
+from kedro.pipeline.node import Node as KedroNode
+
+from kedro_viz.constants import VIZ_METADATA_ARGS
+from kedro_viz.launchers.utils import _find_kedro_project
+from kedro_viz.utils import _hash, _hash_input_output
+
+logger = logging.getLogger(__name__)
+
+TIME_FORMAT = "%Y-%m-%dT%H.%M.%S.%fZ"
+EVENTS_DIR = VIZ_METADATA_ARGS["path"]
+EVENTS_FILE = "kedro_pipeline_events.json"
+
+
+def hash_node(node: Any) -> str:
+    """Stable ID for KedroNode or I/O reference."""
+    return _hash(str(node)) if isinstance(node, KedroNode) else _hash_input_output(node)
+
+
+def create_dataset_event(
+    event_type: str,
+    dataset_name: str,
+    dataset_value: Any = None,
+    datasets: Optional[Dict[str, Any]] = None,
+) -> Dict[str, Any]:
+    """Generic builder for dataset load/save events.
+
+    Args:
+        event_type: Event type/name
+        dataset_name: Dataset name
+        dataset_value: Dataset data
+        datasets: Dictionary of available datasets
+
+    Returns:
+        Dictionary with event data
+    """
+    event: Dict[str, Any] = {
+        "event": event_type,
+        "dataset": dataset_name,
+        "node_id": _hash_input_output(dataset_name),
+        "status": "Available",
+    }
+
+    if dataset_value is not None and datasets:
+        size = compute_size(dataset_name, dataset_value, datasets)
+        if size is not None:
+            event["size_bytes"] = size  # only attach size when available
+    return event
+
+
+def compute_size(
+    dataset_name: str, dataset_value: Any, datasets: Dict[str, Any]
+) -> Optional[int]:
+    """Determine file size for DataFrame or dataset with filepath attribute.
+
+    Args:
+        dataset_name: Dataset name
+        dataset_value: Dataset data
+        datasets: Dictionary of available datasets
+
+    Returns:
+        File size in bytes, if available
+    """
+    dataset = datasets.get(dataset_name)
+    if not dataset:
+        return None
+
+    # pandas DataFrame may store filepath metadata
+    try:
+        import pandas as pd
+
+        if isinstance(dataset_value, pd.DataFrame):
+            file_path = getattr(dataset, "filepath", None) or getattr(
+                dataset, "_filepath", None
+            )
+            if file_path:
+                filesystem, path = fsspec.core.url_to_fs(file_path)
+                return filesystem.size(path) if filesystem.exists(path) else None
```

Comment:
What's the benefit of having separate logic for a pandas dataframe? In the end this is pretty much identical to the below.